consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java

```diff
@@ -0,0 +1,201 @@
+ * under the License.
+ */
+package org.apache.iotdb.consensus.ratis.metrics;
+
+import org.apache.iotdb.metrics.AbstractMetricService;
+import org.apache.iotdb.metrics.utils.MetricLevel;
+import org.apache.iotdb.metrics.utils.MetricType;
+
+import com.codahale.metrics.ConsoleReporter;
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.JmxReporter;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricFilter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.MetricSet;
+import com.codahale.metrics.Timer;
+import org.apache.ratis.metrics.MetricRegistryInfo;
+import org.apache.ratis.metrics.RatisMetricRegistry;
+
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class IoTDBMetricRegistry implements RatisMetricRegistry {
+
+  private final AbstractMetricService metricService;
+  private final MetricRegistryInfo info;
+  private final String prefix;
+  private final Map<String, String> metricNameCache = new ConcurrentHashMap<>();
+  private final Map<String, CounterProxy> counterCache = new ConcurrentHashMap<>();
+  private final Map<String, TimerProxy> timerCache = new ConcurrentHashMap<>();
+
+  private final Map<String, GaugeProxy> gaugeCache = new ConcurrentHashMap<>();
+
+  IoTDBMetricRegistry(MetricRegistryInfo info, AbstractMetricService service) {
+    this.info = info;
+    this.metricService = service;
+    prefix =
+        MetricRegistry.name(
+            info.getApplicationName(), info.getMetricsComponentName(), info.getPrefix());
+  }
+
+  private String getMetricName(String name) {
+    return metricNameCache.computeIfAbsent(name, n -> MetricRegistry.name(prefix, n));
+  }
+
+  @Override
+  public Timer timer(String name) {
+    final String fullName = getMetricName(name);
+    return timerCache.computeIfAbsent(
+        fullName, fn -> new TimerProxy(metricService.getOrCreateTimer(fn, MetricLevel.IMPORTANT)));
+  }
+
+  @Override
+  public Counter counter(String name) {
+    final String fullName = getMetricName(name);
+    return counterCache.computeIfAbsent(
+        fullName,
+        fn ->
+            new CounterProxy(
+                metricService.getOrCreateCounter(getMetricName(name), MetricLevel.IMPORTANT)));
+  }
+
+  @Override
+  public boolean remove(String name) {
+    // TODO (szywilliam) we can add an interface like removeTypeless(name)
+
+    try {
+      metricService.remove(MetricType.COUNTER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.TIMER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.AUTO_GAUGE, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
```

Comment:
If we ignore exceptions here, why use three try?

Reply:
To make each of them independent